cmd/libsnap-confine-private/cgroup-support.c

```diff
@@ -97,3 +103,141 @@ bool sc_cgroup_is_v2() {
     }
     return false;
 }
+
+/**
+ * sc_find_tracking_hierarchy produces the full /sys/fs/cgroup/ path the
+ * selected cgroup hierarchy of the given process. The result is written to
+ * path buffer. If the location cannot be found or any other error occurs, the
+ * process dies.
+ *
+ * The hierarchy selected can be used for process tracking. It is picked based
+ * on the following criteria:
+ *
+ * 1) A unified or v2 cgroup is preferred. If one exists it is selected.
+ * 2) A tracking name=systemd v1 cgroup is supported.
+ **/
+static void sc_find_tracking_hierarchy(pid_t pid, char *path, size_t path_size) {
+    /* Find the path of the unified hierarchy of the given process. */
+    char proc_pid_cgroup[PATH_MAX + 1] = {};
+    sc_must_snprintf(proc_pid_cgroup, sizeof proc_pid_cgroup, "/proc/%d/cgroup", (int)pid);
+    int fd = open(proc_pid_cgroup, O_RDONLY | O_NOFOLLOW | O_CLOEXEC);
+    if (fd < 0) {
+        die("cannot open %s", proc_pid_cgroup);
+    }
+    /* The ownership of the descriptor is now passed to the stream. */
+    FILE *stream = fdopen(fd, "r");
+    if (stream == NULL) {
+        die("cannot open stream from %d", fd);
+    }
+    /* The format of the file is: set of lines representing records. Each
+     * record is a tuple (hierarchy-id,controller-list,cgroup-path) using
+     * colons as element separators. Controllers is in turn a list using commas
+     * as separators. See cgroups(7) for authoritative reference. */
+    char *line = NULL;
+    size_t size = 0;
+    bool found = false;
+    while (true) {
+        /* Read the next line, chomp the trailing newline and parse it. */
+        errno = 0;
+        ssize_t len = getline(&line, &size, stream);
+        if (len < 0) {
+            if (errno != 0) {
+                die("cannot read subsequent line from %s", proc_pid_cgroup);
+            }
+            break;
+        }
+        if (len > 0 && line[len - 1] == '\n') {
+            line[len - 1] = '\0';
+            len--;
+        }
+        char *attr_id, *attr_ctrl_list, *attr_path, *tmp;
+
+        attr_id = line;
+
+        tmp = strchr(attr_id, ':');
+        if (tmp == NULL) {
+            die("cannot parse cgroup, expected hierarchy id: %s", line);
+        }
+        attr_ctrl_list = tmp + 1;
+
+        tmp = strchr(attr_ctrl_list, ':');
+        if (tmp == NULL) {
+            die("cannot parse cgroup, expected controller list: %s", line);
+        }
+        attr_path = tmp + 1;
+
+        tmp = strchr(attr_path, ':');
+        if (tmp != NULL) {
+            die("cannot parse cgroup, expected end of line: %s", line);
+        }
+        attr_ctrl_list[-1] = '\0';
+        attr_path[-1] = '\0';
+
+        debug("cgroup presence: id:%s, controllers:%s, path:%s", attr_id, attr_ctrl_list, attr_path);
+
+        /* Find preferred tracking cgroup, either v1 name=systemd or v2. */
```

Comment:
This comment seems a bit terse for the algorithm used; plus, it doesn't say that v2 is preferred. Perhaps something like:
```
Find a cgroup that can be used for process tracking, preferring
v2 if found, otherwise look for v1 name=systemd cgroup. See
above for more details.
```